packages/server/modules/core/services/ratelimiter.ts

```diff
@@ -0,0 +1,170 @@
+    duration: 3 * TIME.second
+  },
+  ALL_REQUESTS: {
+    limitCount: getIntFromEnv('RATELIMIT_USER_CREATE', '864000'), // 10 per second
+    duration: 1 * TIME.day
+  },
+  USER_CREATE: {
+    limitCount: getIntFromEnv('RATELIMIT_USER_CREATE', '1000'),
+    duration: 1 * TIME.week
+  },
+  STREAM_CREATE: {
+    limitCount: getIntFromEnv('RATELIMIT_STREAM_CREATE', '10000'), // 0.11 per second
+    duration: 1 * TIME.week
+  },
+  COMMIT_CREATE: {
+    limitCount: getIntFromEnv('RATELIMIT_COMMIT_CREATE', '86400'), // 1 per second
+    duration: 1 * TIME.day
+  },
+  'POST /api/getobjects/:streamId': {
+    limitCount: getIntFromEnv('RATELIMIT_POST_GETOBJECTS_STREAMID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'POST /api/diff/:streamId': {
+    limitCount: getIntFromEnv('RATELIMIT_POST_DIFF_STREAMID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'POST /objects/:streamId': {
+    limitCount: getIntFromEnv('RATELIMIT_POST_OBJECTS_STREAMID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'GET /objects/:streamId/:objectId': {
+    limitCount: getIntFromEnv('RATELIMIT_GET_OBJECTS_STREAMID_OBJECTID', '200'),
+    duration: 1 * TIME.minute
+  },
+  'GET /objects/:streamId/:objectId/single': {
+    limitCount: getIntFromEnv('RATELIMIT_GET_OBJECTS_STREAMID_OBJECTID_SINGLE', '200'),
+    duration: 1 * TIME.minute
+  }
+}
+
+const redisClient = new Redis(getRedisUrl(), {
+  enableReadyCheck: false,
+  maxRetriesPerRequest: null
+})
+
+export const sendRateLimitResponse = (
+  res: express.Response,
+  action: string,
+  rateLimiterRes: RateLimiterRes | undefined
+): express.Response => {
+  if (rateLimiterRes instanceof Error) {
+    Sentry.captureException(rateLimiterRes)
+    res.setHeader('X-Speckle-Meditation', 'https://http.cat/500')
+    return res.status(500).send({
+      err: 'Error when attempting to determine rate limit. Please try again later.'
+    })
+  }
+  if (rateLimiterRes) {
+    res.setHeader('Retry-After', rateLimiterRes.msBeforeNext / 1000)
+    res.setHeader('X-RateLimit-Remaining', rateLimiterRes.remainingPoints)
+    res.setHeader(
+      'X-RateLimit-Reset',
+      new Date(Date.now() + rateLimiterRes.msBeforeNext).toISOString()
+    )
+  }
+  if (action) {
+    const opts = LIMITS[action]
+    res.setHeader('X-RateLimit-Limit', opts.limitCount)
+  }
+
+  res.setHeader('X-Speckle-Meditation', 'https://http.cat/429')
+  return res.status(429).send({
+    err: 'You are sending too many requests. You have been rate limited. Please try again later.'
+  }) // TODO we should return a branded page (either here, or via nginx)
+}
+
+export const rateLimiterMiddleware = async (
+  req: express.Request,
+  res: express.Response,
+  next: express.NextFunction
+) => {
+  if (isTestEnv()) next()
+
+  const burstAction = 'ALL_REQUESTS_BURST'
+  const action = 'ALL_REQUESTS'
+
+  const source: string = (req as RequestWithContext)?.context?.userId
+    ? ((req as RequestWithContext)?.context?.userId as string)
```

Comment:
its going to be a problem, because the ratelimiter middleware is now called before the context middleware and the context doesn't exist on the req object when ratelimiting is evaluated. I'll look into this.